include/picongpu/fields/background/templates/twtsfast/BField.tpp

```diff
@@ -0,0 +1,760 @@
+                    / (float_T(2.0) * cspeed * math::pow(helpVar4, float_T(1.5)));
+
+                return result.get_real() / UNIT_SPEED;
+            }
+
+            /** Calculate the Bz(r,t) field
+             *
+             * @param pos Spatial position of the target field.
+             * @param time Absolute time (SI, including all offsets and transformations)
+             *             for calculating the field */
+            HDINLINE
+            BField::float_T BField::calcTWTSBz_Ex(float3_64 const& pos, float_64 const time) const
+            {
+                using complex_T = pmacc::math::Complex<float_T>;
+
+                /* propagation speed of overlap normalized to the speed of light [Default: beta0=1.0] */
+                float_T const beta0 = float_T(beta_0);
+                /* If phi < 0 the formulas below are not directly applicable.
+                 * Instead phi is taken positive, but the entire pulse rotated by 180 deg around the
+                 * z-axis of the coordinate system in this function.
+                 */
+                float_T const phiReal = float_T(math::abs(phi));
+                float_T sinPhiReal;
+                float_T cosPhiReal;
+                pmacc::math::sincos(phiReal, sinPhiReal, cosPhiReal);
+                float_T const alphaTilt = math::atan2(float_T(1.0) - beta0 * cosPhiReal, beta0 * sinPhiReal);
+
+                /* Definition of the laser pulse front tilt angle for the laser field below.
+                 *
+                 * For beta0=1.0, this is equivalent to our standard definition. Question: Why is the
+                 * local "phi_T" not equal in value to the object member "phiReal" or "phi"?
+                 * Because the standard TWTS pulse is defined for beta0 = 1.0 and in the coordinate-system
+                 * of the TWTS model phi is responsible for pulse front tilt and dispersion only. Hence
+                 * the dispersion will (although physically correct) be slightly off the ideal TWTS
+                 * pulse for beta0 != 1.0. This only shows that this TWTS pulse is primarily designed for
+                 * scenarios close to beta0 = 1.
+                 */
+                float_T const phiT = float_T(2.0) * alphaTilt;
+
+                /* Angle between the laser pulse front and the y-axis.
+                 * Not used, but remains in code for documentation purposes.
+                 * float_T const eta = float_T(float_T(PI / 2)) - (phiReal - alphaTilt);
+                 */
+
+                float_T const cspeed = float_T(SI::SPEED_OF_LIGHT_SI / UNIT_SPEED);
+                float_T const lambda0 = float_T(wavelength_SI / UNIT_LENGTH);
+                float_T const om0 = float_T(2.0 * PI * cspeed / lambda0);
+                /* factor 2  in tauG arises from definition convention in laser formula */
+                float_T const tauG = float_T(pulselength_SI * 2.0 / UNIT_TIME);
```

Comment:
please check the precision ` * 2.0` will be an upcast from `float_T` to `float_64`. The division will then be performed in 64bit.

Reply:
All quantities here are float_64, so that depending on the definition of the requested precision of float_T, we have one down-cast or no cast at all. Alternatively I could wrap every factor in its own float_T bracket.

Reply:
Yes you are right!